csrc/lower_thread_predicate.cpp

```diff
@@ -494,6 +519,224 @@ class RedundantUseAnalysis : BackwardVisitor {
+                                              .getDisjointSetOf(rd);
+        for (auto mapped_rd : mapped_rd_exact_set) {
+          std::unordered_set<IterDomain*> all_cids =
+              GpuLower::current()
+                  ->concretizedBroadcastDomains()
+                  ->allConcretizedDomains(mapped_rd);
+          if (!all_cids.empty()) {
+            concretized_broadcast_root_domains[rd] = *all_cids.begin();
+            break;
+          }
+        }
+      }
+    }
+    return concretized_broadcast_root_domains;
+  };
+
+  // Find all the root domains that are merged to the leaf domain.
+  // e.g. Root: [I1,B2,B3] -> Leaf: [I1B2B3], the merged_root_domains =
+  // {B3,B2,I1}
+  auto getRootDomainsMergedToLeaf = [&](IterDomain* ld) {
+    std::vector<IterDomain*> merged_root_domains;
+    std::vector<int> index_root_domain;
+    std::vector<IterDomain*> intermediate_domains = root_domain;
+    auto all_exp = DependencyCheck::getAllExprsBetween(
+        {root_domain.begin(), root_domain.end()}, {ld});
+    for (auto expr : all_exp) {
+      if (auto merge = dynamic_cast<Merge*>(expr)) {
+        auto outer_iter_im = std::find(
+            intermediate_domains.begin(),
+            intermediate_domains.end(),
+            merge->outer());
+        auto inner_iter_im = std::find(
+            intermediate_domains.begin(),
+            intermediate_domains.end(),
+            merge->inner());
+        TORCH_INTERNAL_ASSERT(
+            outer_iter_im != intermediate_domains.end(),
+            "Couldn't find ",
+            merge->outer());
+        TORCH_INTERNAL_ASSERT(
+            inner_iter_im != intermediate_domains.end(),
+            "Couldn't find ",
+            merge->inner());
+        int dist = std::distance(inner_iter_im, outer_iter_im);
+        if (std::abs(dist) != 1) {
+          // if the merged domains are not adjacent, return empty so the
+          // predication is not applied.
+          return std::vector<IterDomain*>();
+        }
+        // insert after erase so the beg iter is not invalidated
+        auto beg = dist > 0 ? inner_iter_im : outer_iter_im;
+        auto end = dist > 0 ? outer_iter_im : inner_iter_im;
+        intermediate_domains.erase(beg, end + 1);
+        intermediate_domains.insert(beg, merge->out());
+
+        auto outer_iter =
+            std::find(root_domain.begin(), root_domain.end(), merge->outer());
+        auto inner_iter =
+            std::find(root_domain.begin(), root_domain.end(), merge->inner());
+        // outer_iter and inner_iter should be neighbors in root_domain
+        if (inner_iter != root_domain.end() &&
+            outer_iter != root_domain.end() &&
+            std::distance(outer_iter, inner_iter) != 1) {
+          return std::vector<IterDomain*>();
+        }
+        if (inner_iter != root_domain.end()) {
+          merged_root_domains.emplace_back(*inner_iter);
+          index_root_domain.emplace_back(inner_iter - root_domain.begin());
```

Comment:
should this and the instance below be `std::distance(root_domain.begin(), inner_iter)`?